massa-client/src/tests/scenarios.rs

```diff
@@ -1,40 +1,51 @@
+use anyhow::{bail, Result};
 use assert_cmd::Command;
 use serial_test::serial;
-use std::thread;
 
-#[tokio::test]
-#[serial]
-async fn test_if_exit_gracefully() {
-    let mut cmd = Command::cargo_bin("massa-client").unwrap();
-    cmd.arg("exit").assert().success();
+const TIMEOUT: u64 = 5;
+const TENTATIVES: u64 = 5;
+
+async fn run_client_cmd(cmd: &str) -> Result<String> {
+    for _ in 0..TENTATIVES {
+        let output = std::str::from_utf8(
+            &Command::new("cargo")
+                .args(&["run", "--", cmd, "--json"])
+                .assert()
+                .get_output()
+                .stdout,
+        )?
+        .to_string();
+        // TODO: compare with tcp connect error as a JSON object?
+        if !output.contains("tcp connect error") {
+            return Ok(output);
+        }
+        tokio::time::sleep(std::time::Duration::from_secs(TIMEOUT)).await;
+    }
+    bail!("tcp connect error")
 }
 
-const CONFIG_PATH: &str = "../massa-node/src/tests/config.toml";
-const TIMEOUT: u64 = 5;
 #[tokio::test]
 #[serial]
-async fn test_if_node_stop() {
-    let massa_node_thread_handle = thread::spawn(|| {
-        Command::cargo_bin("massa-node")
+async fn run_node() {
+    // TODO: @adrien-zinger do we have a better way to spawn `massa_node`?
+    std::thread::spawn(|| {
+        Command::new("cargo")
+            .arg("run")
+            .current_dir("../massa-node")
             .unwrap()
-            .env("MASSA_CONFIG_PATH", CONFIG_PATH)
-            .timeout(std::time::Duration::from_secs(61))
-            .assert()
-            .success();
     });
-    let mut cmd = Command::cargo_bin("massa-client").unwrap();
-    let mut success = false;
-    for _ in 0..5 {
-        tokio::time::sleep(std::time::Duration::from_secs(TIMEOUT)).await;
-        let assert_cli = cmd.arg("node_stop").assert();
-        let output = std::str::from_utf8(&assert_cli.get_output().stdout).unwrap();
-        if !output.contains("tcp connect error") {
-            println!("Client output: {}", output);
-            success = true;
-            break;
-        }
-    }
-    massa_node_thread_handle.join().unwrap();
-    assert!(success, "Error: Failed to close correctly the node 3 times");
-    cmd.arg("exit").assert().success();
+}
+
+#[tokio::test]
+#[serial]
+async fn client_exit_gracefully() {
+    run_client_cmd("exit").await.unwrap();
+}
+
+#[tokio::test]
+#[serial]
+async fn node_stop_gracefully() {
+    run_client_cmd("node_stop").await.unwrap();
+    // TODO: @adrien-zinger do we have a better way than `.join().unwrap()` to
```

Comment:
there is no join anymore

Reply:
Can we join something?

Reply:
I guess this is just an old comment I left and there is nothing to join here. My concern was more to have a review on: did I handle everything right?

Reply:
So you should remove it

Reply:
this test looks ok for me